sdk/javascript/test/utils/arrayHelpers_spec.js

```diff
@@ -0,0 +1,268 @@
+			{ size: 11, tag: 25 },
+			{ size: 12, tag: 36 }
+		];
+
+		it('reads specified number of elements', () => {
+			// Arrange:
+			const context = new TestContext([10, 11, 12, 42, 78]);
+
+			// Act:
+			const elements = arrayHelpers.readArrayCount(context.subView, context.factory, 3);
+
+			// Assert:
+			expect(elements).to.deep.equal(expectedElements);
+		});
+
+		it('can read when using accessor and elements are ordered', () => {
+			// Arrange:
+			const context = new TestContext([10, 11, 12, 42, 78]);
+
+			// Act:
+			const elements = arrayHelpers.readArrayCount(context.subView, context.factory, 3, element => element.tag);
+
+			// Assert:
+			expect(elements).to.deep.equal(expectedElements);
+		});
+
+		it('cannot read when using accessor and elements are not ordered', () => {
+			// Arrange:
+			const context = new TestContext([10, 11, 12, 42, 78]);
+
+			// Assert: use negative to fake wrong ordering
+			expect(() => arrayHelpers.readArrayCount(context.subView, context.factory, 3, element => -element.tag))
+				.to.throw('elements in array are not sorted');
+		});
+
+		assertThrowsWhenAnyElementHasZeroSize((view, factory) => arrayHelpers.readArrayCount(view, factory, 3));
+	});
+
+	describe('readVariableSizeElements', () => {
+		it('reads all available elements', () => {
+			// Arrange: aligned sizes 8, 12, 12, 16, 4
+			const context = new TestContext([7, 11, 12, 13, 3]);
+			const expectedElements = [
+				{ size: 7, tag: 15 },
+				{ size: 11, tag: 15 + 8 },
+				{ size: 12, tag: 15 + 8 + 12 },
+				{ size: 13, tag: 15 + 8 + 12 + 12 },
+				{ size: 3, tag: 15 + 8 + 12 + 12 + 16 }
+			];
+
+			// Act:
+			const elements = arrayHelpers.readVariableSizeElements(context.subView, context.factory, 4);
+
+			// Assert:
+			expect(elements).to.deep.equal(expectedElements);
+		});
+
+		it('throws when reading would result in OOB read', () => {
+			// Arrange:
+			const context1 = new TestContext([24, 25], 49);
+			const context2 = new TestContext([24, 25], 49);
+
+			// Act:
+			const elements = arrayHelpers.readArray(context1.subView, context1.factory);
+
+			// Sanity:
+			expect(elements).to.deep.equal([{ size: 24, tag: 15 }, { size: 25, tag: 15 + 24 }]);
+
+			// Assert:
+			expect(() => arrayHelpers.readVariableSizeElements(context2.subView, context2.factory, 4)).to.throw('unexpected buffer length');
+		});
+
+		assertThrowsWhenAnyElementHasZeroSize((view, factory) => arrayHelpers.readVariableSizeElements(view, factory, 4));
+	});
+
+	class MockElement {
+		constructor(size) {
+			this.size = size;
+		}
+
+		serialize() {
+			return 100 + this.size;
+		}
+	}
+
+	const createElements = () => {
+		const elements = [];
+		for (let i = 0; 5 > i; ++i)
+			elements.push(new MockElement((i * 3) + 1));
+
+		return elements;
+	};
+
+	class WriteCollector {
+		constructor() {
+			this.writes = [];
+		}
+
+		write(obj) {
+			const type = obj instanceof Uint8Array ? 'fill' : 'value';
+			const value = obj instanceof Uint8Array ? obj.length : obj;
```

Comment:
i think it would be clearer with single branch
i.e.
```
this.writes.push(value instanceof Uint8Array ? { type: 'fill', value: value.length } : { type: 'value', value })
```